File: 프로그래머스/2/42586. 기능개발/기능개발.js
// function solution(progresses, speeds) {
//     const days = progresses.map((progress, index) => Math.ceil((100 - progress) / speeds[index]));
//     const result = [0]; 
//     let maxDay = days[0];

//     for (let i = 1, pointer = 0; i < days.length; i++) { 
//         if (days[i] <= maxDay) {
//             result[pointer] += 1; 
//         } else {
//             result[++pointer] = 1; 
//             maxDay = days[pointer];
//         }
//     }

//     return result;
// }

function solution(progresses, speeds) {
    let answer = [0];
    let days = progresses.map((progress, index) => Math.ceil((100 - progress) / speeds[index]));
    let maxDay = days[0];

    for(let i = 0, j = 0; i< days.length; i++){
        if(days[i] <= maxDay) {
            answer[j] += 1;
        } else {
            maxDay = days[i];
            answer[++j] = 1;
        }
    }

    return answer;
}

File: 프로그래머스/2/42586. 기능개발/기능개발.java
import java.util.*;

class Solution {
    public int[] solution(int[] progresses, int[] speeds) {
        List<Integer> answer = new ArrayList<>();
        int[] days = new int[progresses.length];
        
        for (int i = 0; i < progresses.length; i++) {
            days[i] = (100 - progresses[i] + speeds[i] - 1) / speeds[i];
        }
        
        int maxDay = days[0];
        int count = 0;
        
        for (int day : days) {
            if (day <= maxDay) {
                count++;
            } else {
                answer.add(count);
                count = 1;
                maxDay = day;
            }
        }
        answer.add(count);
        
        return answer.stream().mapToInt(i -> i).toArray();
    }
}